flang/lib/Optimizer/Support/InternalNames.cpp

```diff
@@ -164,6 +166,8 @@ std::string
 fir::NameUniquer::doProcedure(llvm::ArrayRef<llvm::StringRef> modules,
                               llvm::Optional<llvm::StringRef> host,
                               llvm::StringRef name) {
+  if (OldABI)
+    return toLower(name)+"_";
```

Comment:
This isn't exactly how we want to achieve this.

Uniqueness is a property we require to avoid collisions in the symbol name space (FIR). Unique names should not be viewed as yet another vendor ABI.

We will want to rewrite the unique names to some selected ABI du jour naming scheme in a designated rewrite pass after high level optimizations. Unfortunately, we haven't written that pass at present.

Reply:
How would this best be achieved? The inability to interoperate with the "standard" abi used by gfortran and classic flang makes using f18 a bit difficult and I'd love to help fix. Regardless, I'll remove the ABI piece for now and get this in for the math functions.

Reply:
The plan is to keep the mangled name at the FIR level, they give the full context. Then, an MLIR pass should be added to mangle the symbols according the desired ABI before LLVM code generation.

There is already some code to demangle the names here (look for `DeconstructedName`): https://github.com/llvm/llvm-project/blob/main/flang/include/flang/Optimizer/Support/InternalNames.h).

The idea would be to add an mlir pass going through the fir::FuncOp, and the operations that may use the names (fir::CallOp and fir::AddressOfOp), and to change the names.

An option could be to do that during the TargetRewrite pass that already exists and is going through these ops (see https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/lib/Optimizer/CodeGen/TargetRewrite.cpp and https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/include/flang/Optimizer/CodeGen/CGPasses.td), or to make it its own pass. I am not sure what would be @schweitzpgi preferred approach here. Both options would make sense to me.

If you want to add have a go at that, that would be great !